### PR TITLE
Internal #2907: Validate Time Zones

### PR DESCRIPTION
**Pull request by hawkfish**
Have SetTimeZoneICU validate the TZ name, just like SET does. Also fix some clangd integer warnings.

fixes: duckdblabs/duckdb-internal#2907

### CODE REVIEW

File: extension/icu/icu-datefunc.cpp [hunk 1]
```diff
@@ -73,6 +73,9 @@ unique_ptr<FunctionData> ICUDateFunc::Bind(ClientContext &context, ScalarFunctio
 
 void ICUDateFunc::SetTimeZone(icu::Calendar *calendar, const string_t &tz_id) {
 	auto tz = icu_66::TimeZone::createTimeZone(icu::UnicodeString::fromUTF8(icu::StringPiece(tz_id.GetString())));
+	if (*tz == icu::TimeZone::getUnknown()) {
+		throw NotImplementedException("Unknown TimeZone '%s'", tz_id.GetString());
```
Comment (carlopi) [inline extension/icu/icu-datefunc.cpp:77]:
```suggestion
		throw InvalidInputException("Unknown TimeZone '%s'", tz_id.GetString());
```
Maybe somewhat better.

**Review by carlopi — APPROVED** [review]
LGTM!

Only minor thing there is a Python test that needs fixing

File: extension/icu/icu-datefunc.cpp (hunk 1 above)
Comment (hawkfish) [reply extension/icu/icu-datefunc.cpp:77]: Ah this is the same exception we are using for `SET TimeZone=` Hard to say whether it is not implemented (they gave us a legal TZ that we don't support) or invalid (they gave us a TZ that doesn't exist anywhere.)

Comment (carlopi) [reply extension/icu/icu-datefunc.cpp:77]: right, cool then

**Comment by Mytherin** [comment]
Looks like the Python test is still failing:

```
../../../project/tests/stubs/test_stubs.py::test_generated_stubs PASSED  [100%]

=================================== FAILURES ===================================
___ TestDateTimeTimeStamp.test_timestamp_timezone_positive_extreme[pandas0] ____

self = <test_datetime_timestamp.TestDateTimeTimeStamp object at 0x7f2574948850>
pandas = <conftest.NumpyPandas object at 0x7f2574beb940>
duckdb_cursor = <duckdb.duckdb.DuckDBPyConnection object at 0x7f241c3f7e30>

    @pytest.mark.skipif(
        Version(pd.__version__) < Version('2.0.2'), reason="pandas < 2.0.2 does not properly convert timezones"
    )
    @pytest.mark.parametrize('pandas', [NumpyPandas(), ArrowPandas()])
    def test_timestamp_timezone_positive_extreme(self, pandas, duckdb_cursor):
        duckdb_time = duckdb_cursor.sql(
            """
            SELECT timestamp '2021-12-31 23:00:00' AT TIME ZONE 'Etc/GMT-14' as "0"
        """
        ).df()
    
        offset = datetime.timedelta(hours=14)
        timezone = datetime.timezone(offset)
    
        df_in = pandas.DataFrame(
            {
                0: pandas.Series(
                    data=[datetime.datetime(year=2022, month=1, day=1, hour=13, tzinfo=timezone)], dtype='object'
                )
            }
        )
        df_out = duckdb_cursor.sql("""select * from df_in""").df()
>       pandas.testing.assert_frame_equal(df_out, duckdb_time)

/project/tests/fast/pandas/test_datetime_timestamp.py:123: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../venv/lib/python3.9/site-packages/pandas/_testing/asserters.py:794: in assert_extension_array_equal
    assert_numpy_array_equal(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

left = array([1640991600000000]), right = array([1640941200000000])
err_msg = None

    def _raise(left, right, err_msg) -> NoReturn:
        if err_msg is None:
            if left.shape != right.shape:
                raise_assert_detail(
                    obj, f"{obj} shapes are different", left.shape, right.shape
                )
    
            diff = 0
            for left_arr, right_arr in zip(left, right):
                # count up differences
                if not array_equivalent(left_arr, right_arr, strict_nan=strict_nan):
                    diff += 1
    
            diff = diff * 100.0 / left.size
            msg = f"{obj} values are different ({np.round(diff, 5)} %)"
>           raise_assert_detail(obj, msg, left, right, index_values=index_values)
E           AssertionError: DataFrame.iloc[:, 0] (column name="0") are different
E           
E           DataFrame.iloc[:, 0] (column name="0") values are different (100.0 %)
E           [index]: [0]
E           [left]:  [1640991600000000]
E           [right]: [1640941200000000]

../venv/lib/python3.9/site-packages/pandas/_testing/asserters.py:684: AssertionError
___ TestDateTimeTimeStamp.test_timestamp_timezone_positive_extreme[pandas1] ____

self = <test_datetime_timestamp.TestDateTimeTimeStamp object at 0x7f25749488b0>
pandas = <conftest.ArrowPandas object at 0x7f2574beb2b0>
duckdb_cursor = <duckdb.duckdb.DuckDBPyConnection object at 0x7f240dcbd3b0>

    @pytest.mark.skipif(
        Version(pd.__version__) < Version('2.0.2'), reason="pandas < 2.0.2 does not properly convert timezones"
    )
    @pytest.mark.parametrize('pandas', [NumpyPandas(), ArrowPandas()])
    def test_timestamp_timezone_positive_extreme(self, pandas, duckdb_cursor):
        duckdb_time = duckdb_cursor.sql(
            """
            SELECT timestamp '2021-12-31 23:00:00' AT TIME ZONE 'Etc/GMT-14' as "0"
        """
        ).df()
    
        offset = datetime.timedelta(hours=14)
        timezone = datetime.timezone(offset)
    
        df_in = pandas.DataFrame(
            {
                0: pandas.Series(
                    data=[datetime.datetime(year=2022, month=1, day=1, hour=13, tzinfo=timezone)], dtype='object'
                )
            }
        )
        df_out = duckdb_cursor.sql("""select * from df_in""").df()
>       pandas.testing.assert_frame_equal(df_out, duckdb_time)

/project/tests/fast/pandas/test_datetime_timestamp.py:123: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/project/tests/conftest.py:155: in convert_and_equal
    import_pandas().testing.assert_frame_equal(df1, df2, **kwargs)
../venv/lib/python3.9/site-packages/pandas/_testing/asserters.py:794: in assert_extension_array_equal
    assert_numpy_array_equal(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

left = array([1640991600000000]), right = array([1640941200000000])
err_msg = None

    def _raise(left, right, err_msg) -> NoReturn:
        if err_msg is None:
            if left.shape != right.shape:
                raise_assert_detail(
                    obj, f"{obj} shapes are different", left.shape, right.shape
                )
    
            diff = 0
            for left_arr, right_arr in zip(left, right):
                # count up differences
                if not array_equivalent(left_arr, right_arr, strict_nan=strict_nan):
                    diff += 1
    
            diff = diff * 100.0 / left.size
            msg = f"{obj} values are different ({np.round(diff, 5)} %)"
>           raise_assert_detail(obj, msg, left, right, index_values=index_values)
E           AssertionError: DataFrame.iloc[:, 0] (column name="0") are different
E           
E           DataFrame.iloc[:, 0] (column name="0") values are different (100.0 %)
E           [index]: [0]
E           [left]:  [1640991600000000]
E           [right]: [1640941200000000]


```

Could you have a look?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**